File: packages/sdk-dotnet/src/Rollgate.SDK/CircuitBreaker.cs
namespace Rollgate.SDK;

public enum CircuitState
{
    Closed,
    Open,
    HalfOpen
}

public class CircuitBreaker
{
    private readonly CircuitBreakerConfig _config;
    private readonly object _lock = new();
    private CircuitState _state = CircuitState.Closed;
    private readonly List<DateTime> _failures = new();
    private DateTime _openedAt;
    private int _halfOpenSuccesses;
    private Action<CircuitState, CircuitState>? _onStateChange;

    public CircuitBreaker(CircuitBreakerConfig config)
    {
        _config = config;
    }

    public void Execute(Action fn)
    {
        if (!IsAllowingRequests())
            throw RollgateException.CircuitOpenError();

        lock (_lock)
        {
            if (_state == CircuitState.Open)
                TransitionTo(CircuitState.HalfOpen);
        }

        try
        {
            fn();
            lock (_lock) { RecordSuccess(); }
        }
        catch
        {
            lock (_lock) { RecordFailure(); }
            throw;
        }
    }

    public async Task ExecuteAsync(Func<Task> fn)
    {
        if (!IsAllowingRequests())
            throw RollgateException.CircuitOpenError();

        lock (_lock)
        {
            if (_state == CircuitState.Open)
                TransitionTo(CircuitState.HalfOpen);
        }

        try
        {
            await fn();
            lock (_lock) { RecordSuccess(); }
        }
        catch
        {
            lock (_lock) { RecordFailure(); }
            throw;
        }
    }

    public bool IsAllowingRequests()
    {
        lock (_lock)
        {
            return _state switch
            {
                CircuitState.Closed => true,
                CircuitState.HalfOpen => true,
                CircuitState.Open => (DateTime.UtcNow - _openedAt) >= _config.RecoveryTimeout,
                _ => true
            };
        }
    }

    public CircuitState GetState()
    {
        lock (_lock) { return _state; }
    }

    public string GetStateString()
    {
        lock (_lock)
        {
            return _state switch
            {
                CircuitState.Closed => "closed",
                CircuitState.Open => "open",
                CircuitState.HalfOpen => "half_open",
                _ => "closed"
            };
        }
    }

    public void OnStateChange(Action<CircuitState, CircuitState> fn)
    {
        lock (_lock) { _onStateChange = fn; }
    }

    private void RecordFailure()
    {
        var now = DateTime.UtcNow;
        _failures.Add(now);
        CleanOldFailures();

        if (_state == CircuitState.HalfOpen)
        {
            TransitionTo(CircuitState.Open);
            return;
        }

        if (CountRecentFailures() >= _config.FailureThreshold)
            TransitionTo(CircuitState.Open);
    }

    private void RecordSuccess()
    {
        if (_state == CircuitState.HalfOpen)
        {
            _halfOpenSuccesses++;
            if (_halfOpenSuccesses >= _config.SuccessThreshold)
                TransitionTo(CircuitState.Closed);
        }
    }

    private void TransitionTo(CircuitState newState)
    {
        if (_state == newState) return;
        var oldState = _state;
        _state = newState;

        if (newState == CircuitState.Open)
            _openedAt = DateTime.UtcNow;
        if (newState == CircuitState.Closed)
        {
            _failures.Clear();
            _halfOpenSuccesses = 0;
        }
        if (newState == CircuitState.HalfOpen)
            _halfOpenSuccesses = 0;

        _onStateChange?.Invoke(oldState, newState);
    }

    private void CleanOldFailures()
    {
        var cutoff = DateTime.UtcNow - _config.MonitoringWindow;
        _failures.RemoveAll(f => f < cutoff);
    }

    private int CountRecentFailures()
    {
        var cutoff = DateTime.UtcNow - _config.MonitoringWindow;
        return _failures.Count(f => f >= cutoff);
    }
}

File: packages/sdk-flutter/lib/src/circuit_breaker.dart
import 'config.dart';
import 'errors.dart';

enum CircuitState { closed, open, halfOpen }

class CircuitBreaker {
  final CircuitBreakerConfig _config;
  CircuitState _state = CircuitState.closed;
  final List<DateTime> _failures = [];
  DateTime _openedAt = DateTime.now();
  int _halfOpenSuccesses = 0;
  void Function(CircuitState from, CircuitState to)? _onStateChange;

  CircuitBreaker(this._config);

  Future<void> execute(Future<void> Function() fn) async {
    if (!isAllowingRequests()) {
      throw RollgateException.circuitOpenError();
    }

    if (_state == CircuitState.open) {
      _transitionTo(CircuitState.halfOpen);
    }

    try {
      await fn();
      _recordSuccess();
    } catch (e) {
      _recordFailure();
      rethrow;
    }
  }

  bool isAllowingRequests() {
    switch (_state) {
      case CircuitState.closed:
      case CircuitState.halfOpen:
        return true;
      case CircuitState.open:
        return DateTime.now().difference(_openedAt) >= _config.recoveryTimeout;
    }
  }

  CircuitState getState() => _state;

  String getStateString() {
    switch (_state) {
      case CircuitState.closed:
        return 'closed';
      case CircuitState.open:
        return 'open';
      case CircuitState.halfOpen:
        return 'half_open';
    }
  }

  void onStateChange(void Function(CircuitState, CircuitState) fn) {
    _onStateChange = fn;
  }

  void _recordFailure() {
    final now = DateTime.now();
    _failures.add(now);
    _cleanOldFailures();

    if (_state == CircuitState.halfOpen) {
      _transitionTo(CircuitState.open);
      return;
    }

    if (_countRecentFailures() >= _config.failureThreshold) {
      _transitionTo(CircuitState.open);
    }
  }

  void _recordSuccess() {
    if (_state == CircuitState.halfOpen) {
      _halfOpenSuccesses++;
      if (_halfOpenSuccesses >= _config.successThreshold) {
        _transitionTo(CircuitState.closed);
      }
    }
  }

  void _transitionTo(CircuitState newState) {
    if (_state == newState) return;
    final oldState = _state;
    _state = newState;

    if (newState == CircuitState.open) _openedAt = DateTime.now();
    if (newState == CircuitState.closed) {
      _failures.clear();
      _halfOpenSuccesses = 0;
    }
    if (newState == CircuitState.halfOpen) _halfOpenSuccesses = 0;

    _onStateChange?.call(oldState, newState);
  }

  void _cleanOldFailures() {
    final cutoff = DateTime.now().subtract(_config.monitoringWindow);
    _failures.removeWhere((f) => f.isBefore(cutoff));
  }

  int _countRecentFailures() {
    final cutoff = DateTime.now().subtract(_config.monitoringWindow);
    return _failures.where((f) => f.isAfter(cutoff)).length;
  }
}
